core/routing/routing.go

```diff
@@ -36,6 +37,16 @@ type ContentRouting interface {
 	FindProvidersAsync(context.Context, cid.Cid, int) <-chan peer.AddrInfo
 }
 
+// ContentRoutingMany complements ContentRouting intetface allowing providing several keys at
+// the same time, improving performance.
+type ContentRoutingMany interface {
+	// ProvideMany adds the given keys to the content routing system.
+	ProvideMany(ctx context.Context, keys []multihash.Multihash) error
```

Comment:
Why does the signature look so different from `ContentRouting.Provide`? I would have expected it to be identical to `Provide`, but just take a slice of CIDs.

Reply:
I think it is historically coming from the FullRT DHT implementation: https://github.com/libp2p/go-libp2p-kad-dht/blob/master/fullrt/dht.go#L914

Reply:
@ajnavarro some historical context. I used this signature for the DHT because it really only advertises multihashes and it seemed misleading (and incorrect) to use CIDs in the signature, also the `broadcast` boolean seemed very out of place ... why call Provide if you're not trying to broadcast?

That being said if it's useful to change `ProvideMany`'s signature to something else (e.g. taking a slice of CIDs, taking a channel of CIDs, taking an iterator that yields CIDs, ...) that seems pretty reasonable.

Given that recently merged Provide methods for Reframe (https://github.com/ipfs/specs/blob/04dbe0267edbbc2c9a3a9df3540e71327fabf86d/reframe/REFRAME_KNOWN_METHODS.md#provide) take CIDs it might be more natural to switch to providing CIDs. This does mean taking a look at the DHT code to make sure it's not wasting a ton of memory as you do the conversion though and might help you figure out which signature is correct.